src/org/traccar/reports/Events.java

```diff
@@ -47,16 +67,69 @@ public static String getJson(long userId, Collection<Long> deviceIds, Collection
         return json.build().toString();
     }
 
-    public static String getCsv(long userId, Collection<Long> deviceIds, Collection<Long> groupIds,
-            Collection<String> types, Date from, Date to) throws SQLException {
-        CsvBuilder csv = new CsvBuilder();
-        csv.addHeaderLine(new Event());
+    public static void getExcel(OutputStream outputStream,
+            long userId, Collection<Long> deviceIds, Collection<Long> groupIds,
+            Collection<String> types, Date from, Date to) throws SQLException, IOException {
+        ArrayList<DeviceReport> devicesEvents = new ArrayList<>();
+        ArrayList<String> sheetNames = new ArrayList<>();
+        HashMap<Long, String> geofenceNames = new HashMap<>();
         for (long deviceId: ReportUtils.getDeviceList(deviceIds, groupIds)) {
             Context.getPermissionsManager().checkDevice(userId, deviceId);
+            SortedSet<Event> eventReports = new TreeSet<Event>(new Comparator<Event>() {
+                @Override
+                public int compare(Event e1, Event e2) {
+                    return e1.getServerTime().compareTo(e2.getServerTime());
+                }
+            });
             for (String type : types) {
-                csv.addArray(Context.getDataManager().getEvents(deviceId, type, from, to));
+                Collection<Event> events = Context.getDataManager().getEvents(deviceId, type, from, to);
+                for (Event event : events) {
+                    long geofenceId = event.getGeofenceId();
+                    if (geofenceId != 0) {
+                        Geofence geofence = Context.getGeofenceManager().getGeofence(geofenceId);
+                        if (geofence != null) {
+                            geofenceNames.put(geofenceId, geofence.getName());
+                        } else {
+                            geofenceNames.put(geofenceId, Long.toString(geofenceId));
+                        }
+                    }
+                    eventReports.add(event);
+                }
+            }
+            if (!eventReports.isEmpty()) {
+                DeviceReport deviceEvents = new DeviceReport();
+                Device device = Context.getIdentityManager().getDeviceById(deviceId);
+                deviceEvents.setDeviceName(device.getName());
+                sheetNames.add(deviceEvents.getDeviceName());
+                if (device.getGroupId() != 0) {
+                    Group group = Context.getDeviceManager().getGroupById(device.getGroupId());
+                    if (group != null) {
+                        deviceEvents.setGroupName(group.getName());
+                    }
+                }
+                deviceEvents.setObjects(eventReports);
+                devicesEvents.add(deviceEvents);
+            }
+        }
+        String templatePath = Context.getConfig().getString("report.events.template.excel",
```

Comment:
A couple of things here:

1. I don't think we need two subfolders for templates. Just put xlsx files into the "templates" folder.
2. I think it would be easier if in the config we just needed to configure path to the temples, not every file separately.